ONSdigital/eq-questionnaire-runner
# Create supplementary data store

### What is the context of this PR?
This PR implements the creation of the new supplementary data store and the loading of sds data into that store whenever a new questionnaire session is launched (and any existing supplementary dataset id doesn't match)

It handles removing old supplementary data if the new data is different (or there is no new data) and creating or updating the list store for any supplementary data lists.

This is part of the **Loading of supplementary data** card and implements the functionality, along with a simple test for the store, but no other testing yet.

### Implementation decisions
The supplementary data, ideally will be fetched in `session.py` however this needs to happen between the store being created/fetched from storage and the metadata for it being overridden, which all happens in the authenticator. Since it needs the metadata from the login function, a context seemed like the way to go

### How to review
Read through the trello card to understand the requirements and check that the structure of the supplementary data store and how it will be interacted with is suitable 

### How to test functionally
Following the guidance in run-mock-sds-endpoint.md start the mock sds api and pick the `test_supplementary_data` schema. 

1. Save the generated response id somewhere
2. Use the second dataset id from the read-me for the sds dataset id (that one has lists)
3. put a breakpoint on the final line of the login endpoint in `session.py`
4. On opening the survey and hitting that breakpoint, the newly created questionnaire store should have the correct supplementary data, and the list store should also contain the "products" list
5. Using another breakpoint on the first line of `_set_questionnaire_supplementary_data` if you relaunch the survey, with the same response id as before, but using the other supplementary dataset id
    * at first breakpoint the questionnaire store from before is loaded, so will have the old supplementary data, and a list store with "products"
    * on hitting the second breakpoint, the update will be complete, the list store will be empty, and the supplementary data will match that of `supplementary_data_no_repeat.json`
6. Use a break point on `return data` in `converter_v2.py` to check that the supplementary data and list mappings are included correctly.

### Whats next
Tests to cover supplementary data being added, removed, updated on session start. 
Tests for the data being send downstream
Tests to cover a list collector using the supplementary data, and removal of answers from the answer store when supplementary data is removed/updated
